docs/batch jobs/ISEL timetable extraction batch job/isel-timetable-batch-job-architecture.md

```diff
@@ -0,0 +1,87 @@
+![timetable-extraction-flow](isel-timetable-extraction-flow.png)
+
+## Job Architecture
+With Spring Batch we can configure a Job as a sequence of ordered steps. A step can consist of a tasklet (a more flexible piece with no necessity of configuring readers and writers) or a step based on chunks wich will read and process a fixed number of items (a chunk) and write it once. It has associated mandatory ItemReader and ItemWriter definitions and an optional ItemProcessor.
+
+The following diagram presents the sequence of steps included in the ISEL timetable extraction job:
+
+![dependence](isel-timetable-extraction-architecture.png)
+
+Step 1 makes sure that that the timetable document is reachable in the URL specified in the configuration document (for more information on the document, visit [this]() page), and that it wasn't parsed in the past.
+
+The second step verifies that the format of the file is as expected. This is relevant since the extraction process strongly depends on it.
+
+Next, step 3 maps the verified object to a business object, more suited to the upload.
+
+The fourth step is comprised of two tasklets that are executed in parallel, which upload timetable and faculty information to I-On Core.
+
+The final step updates the database with the hash of the parsed file and notifies watchers of success.
+
+Following is a more detailed explanation of what each step does, how it is done, and why.
+
+### Step 1 - Download and Compare
+Given that the timetable extraction job is dependent on information that is not controlled by the I-On Integration project, namely the ISEL timetable PDF, we need to make sure the document is in the designated location. Also, we don't want to send information that is already present in I-On Core. This step downloads the pdf document, (most likely from ISEL's website, but the location is configurable through the configuration document) and writes it to the local filesystem. It calculates a hash of the file. Then it reads the value of the hash of the document used in the last time the job ran successfully. Assuming a non-broken cryptographic hash and no collisions, if the two hashes are the same, then we know that the document content is certainly the same. In that case, the job does not proceed, as the extracted information is already present in I-On Core. On the other hand, if the saved hash is different from the one just calculated, then the file contents have changed since the last run of the job. Because the integration project doesn't save state, this is a fairly inexpensive way of avoiding processing repeated information. It is suited to this document in particular, since it is not expected that it should change frequently. It is normally posted in the start of the semester. There may be an isolated change in the following weeks, but then the document persists until the end of the semester.
```

Comment:
- Break this sentence into multiple lines.
- Do we need to store the downloaded document in the file system?
- Where is the hash saved?

Reply:
We have to store it since [tabula](https://github.com/tabulapdf/tabula-java) which we'll use to obtain the table data, only accepts `File` as input.
The hash will be saved in a postgres database.